File: conformance/driver.py
"""The Python leg of the differential harness. See README.md.

Prints one `label|result` line per case. Every port prints the same labels in
the same order; conformance/compare.py runs all four and diffs.
"""

import os
import sys
from pathlib import Path

# The source next door by default. `compare.py --released` points this at a
# directory it has installed the published wheel into, so the same driver can
# be asked about either without a second copy of it existing.
SOURCE = os.environ.get("GPC_PYTHON_PATH") or str(
    Path(__file__).resolve().parent.parent / "python" / "src"
)
sys.path.insert(0, SOURCE)

from gridpointcode_algo_pranavpatel_ca import GPC  # noqa: E402
from gridpointcode_algo_pranavpatel_ca.errors import GPCError

out = []


def t(label, fn):
    try:
        r = fn()
        if isinstance(r, tuple):
            r = ",".join(fmt(x) for x in r)
        elif isinstance(r, list):
            r = " ".join(str(x) for x in r)
        else:
            r = fmt(r)
        out.append(f"{label}|{r}")
    except GPCError as e:
        out.append(f"{label}|ERR:{e.reason}")
    except Exception as e:                      # noqa: BLE001
        out.append(f"{label}|EXC:{type(e).__name__}")


def fmt(x):
    if isinstance(x, bool):
        return "true" if x else "false"
    if isinstance(x, float):
        return repr(x)
    if isinstance(x, list):
        return " ".join(str(i) for i in x)
    return str(x)


C = "G3RJM98NM9"
X = "XG3RJ98NM9"

# --- cells and levels
for k in (-1, 0, 1, 5, 10, 11):
    t(f"cell({k})", lambda k=k: GPC.cell(C, k))
t("cell(reserved,3)", lambda: GPC.cell(X, 3))
t("cell(lowercase,4)", lambda: GPC.cell("g3rjm98nm9", 4))
t("cell(formatted,4)", lambda: GPC.cell("#G3RJM-98NM9", 4))

# --- contains
t("contains(cell,code)", lambda: GPC.contains("G3RJM", C))
t("contains(self)", lambda: GPC.contains(C, C))
t("contains(other)", lambda: GPC.contains("G3RJM", "G3RJT98NM9"))
t("contains(longer-cell)", lambda: GPC.contains(C, "G3RJM"))
t("contains(reserved-cell)", lambda: GPC.contains("XG3RJ", X))
t("contains(empty-cell)", lambda: GPC.contains("", C))

# --- neighbours
for code in ("G3RJM98NM9", "P4444PPPPP", "3PPPP00000", "F0000000000"[:10], "G", "G3RJM"):
    t(f"neighbours({code})", lambda code=code: GPC.neighbours(code))
t("neighbours(reserved)", lambda: GPC.neighbours(X))

# --- distance
t("distance(self)", lambda: GPC.distance(C, C))
t("distance(levels)", lambda: GPC.distance("G3RJM", C))
t("distance(antipodal)", lambda: GPC.distance("P4444PPPPP", "3PPPP00000"))
t("distance(reserved)", lambda: GPC.distance(X, C))

# --- cell dimensions
for k in (0, 1, 10, 11):
    t(f"cellDimensions({k})", lambda k=k: GPC.cell_dimensions(k))

# --- integer form
t("toInteger", lambda: GPC.to_integer(C))
t("toInteger(reserved)", lambda: GPC.to_integer(X))
t("fromInteger(0)", lambda: GPC.from_integer(0))
t("fromInteger(-1)", lambda: GPC.from_integer(-1))
t("fromInteger(max)", lambda: GPC.from_integer(25 ** 10 - 1))
t("fromInteger(over)", lambda: GPC.from_integer(25 ** 10))

# --- short form
t("shorten", lambda: GPC.shorten(C))
t("recoverShort(dash)", lambda: GPC.recover_short("-98NM9", 43.65, -79.38))
t("recoverShort(plain)", lambda: GPC.recover_short("98NM9", 43.65, -79.38))
t("recoverShort(lower)", lambda: GPC.recover_short("98nm9", 43.65, -79.38))
t("recoverShort(alias)", lambda: GPC.recover_short("98NMO", 43.65, -79.38))
t("recoverShort(short)", lambda: GPC.recover_short("98NM", 43.65, -79.38))
t("recoverShort(long)", lambda: GPC.recover_short("98NM99", 43.65, -79.38))
t("recoverShort(wrap-w)", lambda: GPC.recover_short("00000", 0.0, -179.999))
t("recoverShort(wrap-e)", lambda: GPC.recover_short("00000", 0.0, 179.999))
t("recoverShort(pole)", lambda: GPC.recover_short("PPPPP", 90.0, 0.0))

# --- corrections
t("suggest(level0)", lambda: len(GPC.suggest_corrections(C, 43.65, -79.38, 0)))
t("suggest(level11)", lambda: len(GPC.suggest_corrections(C, 43.65, -79.38, 11)))
t("suggest(reserved-in)", lambda: len(GPC.suggest_corrections(X, 43.65, -79.38)))
t("suggest(bad-char)", lambda: len(GPC.suggest_corrections("G3RJM98NMQ", 43.65, -79.38)))
t("suggest(count6)", lambda: len(GPC.suggest_corrections("G3RJM98N09", 43.65, -79.38, 6)))

# --- screen
t("screen(code)", lambda: GPC.screen(C))
t("screen(reserved)", lambda: GPC.screen(X))
t("screen(bad-char)", lambda: GPC.screen("G3RJM98NMQ"))
t("screen(short)", lambda: GPC.screen("G3RJM"))

# --- conversions
t("toDMS", lambda: GPC.to_dms(43.650006, -79.380004))
t("toDMS(zero)", lambda: GPC.to_dms(0.0, 0.0))
t("toDMS(negzero)", lambda: GPC.to_dms(-0.0, -0.0))
t("toDMS(pole)", lambda: GPC.to_dms(90.0, 180.0))
t("fromDMS", lambda: GPC.from_dms("43\u00b039'00.02\"N, 79\u00b022'48.01\"W"))
t("fromDMS(min60)", lambda: GPC.from_dms("43\u00b060'00.00\"N, 79\u00b022'48.01\"W"))
t("fromDMS(sec60)", lambda: GPC.from_dms("43\u00b039'60.00\"N, 79\u00b022'48.01\"W"))
t("fromDMS(junk)", lambda: GPC.from_dms("hello"))
t("fromDMS(empty)", lambda: GPC.from_dms(""))
t("toGeoURI", lambda: GPC.to_geo_uri(43.650006, -79.380004))
t("toGeoURI(zero)", lambda: GPC.to_geo_uri(0.0, 0.0))
t("fromGeoURI", lambda: GPC.from_geo_uri("geo:43.650006,-79.380004"))
t("fromGeoURI(crs)", lambda: GPC.from_geo_uri("geo:43.65,-79.38;crs=WGS84"))
t("fromGeoURI(badcrs)", lambda: GPC.from_geo_uri("geo:43.65,-79.38;crs=nad27"))
t("fromGeoURI(u)", lambda: GPC.from_geo_uri("geo:43.65,-79.38;u=35"))
t("fromGeoURI(alt)", lambda: GPC.from_geo_uri("geo:43.65,-79.38,120"))
t("fromGeoURI(junk)", lambda: GPC.from_geo_uri("http://x"))
t("fromGeoURI(empty)", lambda: GPC.from_geo_uri(""))

# --- bulk
t("encodeAll(empty)", lambda: len(GPC.encode_all([])))
t("decodeAll(empty)", lambda: len(GPC.decode_all([])))
t("encodeAll(two)", lambda: GPC.encode_all([(43.65, -79.38), (0.0, 0.0)]))

# --- check form
t("withCheck", lambda: GPC.with_check(C))
t("withCheck(reserved)", lambda: GPC.with_check(X))

print("\n".join(out))

File: conformance/driver.js
// The TypeScript leg of the differential harness. See README.md.
// Reads the built package, so `npm run build` in typescript/ has to have run.
//
// `compare.py --released` points GPC_TYPESCRIPT_MAIN at a directory it has
// installed the published package into, so the same driver can be asked about
// either without a second copy of it existing.
const path = require('path');
const main = process.env.GPC_TYPESCRIPT_MAIN
    || path.join(__dirname, '..', 'typescript', 'dist', 'index.js');
const { GPC } = require(main);

const out = [];
function fmt(x) {
    if (typeof x === 'boolean') return x ? 'true' : 'false';
    if (Array.isArray(x)) return x.map(fmt).join(' ');
    if (typeof x === 'number') return Number.isInteger(x) && Math.abs(x) < 1e15 ? String(x) : String(x);
    return String(x);
}
function t(label, fn) {
    try {
        const r = fn();
        let s;
        if (Array.isArray(r)) {
            // tuple-ish results are comma joined, list results space joined
            s = r.every((e) => typeof e === 'string') && label.startsWith('neighbours')
                ? r.join(' ')
                : r.map(fmt).join(',');
        } else s = fmt(r);
        out.push(`${label}|${s}`);
    } catch (e) {
        if (e && e.reason) out.push(`${label}|ERR:${e.reason}`);
        else out.push(`${label}|EXC:${e && e.constructor ? e.constructor.name : 'x'}`);
    }
}

const C = 'G3RJM98NM9';
const X = 'XG3RJ98NM9';

for (const k of [-1, 0, 1, 5, 10, 11]) t(`cell(${k})`, () => GPC.cell(C, k));
t('cell(reserved,3)', () => GPC.cell(X, 3));
t('cell(lowercase,4)', () => GPC.cell('g3rjm98nm9', 4));
t('cell(formatted,4)', () => GPC.cell('#G3RJM-98NM9', 4));

t('contains(cell,code)', () => GPC.contains('G3RJM', C));
t('contains(self)', () => GPC.contains(C, C));
t('contains(other)', () => GPC.contains('G3RJM', 'G3RJT98NM9'));
t('contains(longer-cell)', () => GPC.contains(C, 'G3RJM'));
t('contains(reserved-cell)', () => GPC.contains('XG3RJ', X));
t('contains(empty-cell)', () => GPC.contains('', C));

for (const code of ['G3RJM98NM9', 'P4444PPPPP', '3PPPP00000', 'F000000000', 'G', 'G3RJM'])
    t(`neighbours(${code})`, () => GPC.neighbours(code));
t('neighbours(reserved)', () => GPC.neighbours(X));

t('distance(self)', () => GPC.distance(C, C));
t('distance(levels)', () => GPC.distance('G3RJM', C));
t('distance(antipodal)', () => GPC.distance('P4444PPPPP', '3PPPP00000'));
t('distance(reserved)', () => GPC.distance(X, C));

for (const k of [0, 1, 10, 11]) t(`cellDimensions(${k})`, () => GPC.cellDimensions(k));

t('toInteger', () => GPC.toInteger(C));
t('toInteger(reserved)', () => GPC.toInteger(X));
t('fromInteger(0)', () => GPC.fromInteger(0));
t('fromInteger(-1)', () => GPC.fromInteger(-1));
t('fromInteger(max)', () => GPC.fromInteger(25 ** 10 - 1));
t('fromInteger(over)', () => GPC.fromInteger(25 ** 10));

t('shorten', () => GPC.shorten(C));
t('recoverShort(dash)', () => GPC.recoverShort('-98NM9', 43.65, -79.38));
t('recoverShort(plain)', () => GPC.recoverShort('98NM9', 43.65, -79.38));
t('recoverShort(lower)', () => GPC.recoverShort('98nm9', 43.65, -79.38));
t('recoverShort(alias)', () => GPC.recoverShort('98NMO', 43.65, -79.38));
t('recoverShort(short)', () => GPC.recoverShort('98NM', 43.65, -79.38));
t('recoverShort(long)', () => GPC.recoverShort('98NM99', 43.65, -79.38));
t('recoverShort(wrap-w)', () => GPC.recoverShort('00000', 0.0, -179.999));
t('recoverShort(wrap-e)', () => GPC.recoverShort('00000', 0.0, 179.999));
t('recoverShort(pole)', () => GPC.recoverShort('PPPPP', 90.0, 0.0));

t('suggest(level0)', () => GPC.suggestCorrections(C, 43.65, -79.38, 0).length);
t('suggest(level11)', () => GPC.suggestCorrections(C, 43.65, -79.38, 11).length);
t('suggest(reserved-in)', () => GPC.suggestCorrections(X, 43.65, -79.38).length);
t('suggest(bad-char)', () => GPC.suggestCorrections('G3RJM98NMQ', 43.65, -79.38).length);
t('suggest(count6)', () => GPC.suggestCorrections('G3RJM98N09', 43.65, -79.38, 6).length);

t('screen(code)', () => GPC.screen(C));
t('screen(reserved)', () => GPC.screen(X));
t('screen(bad-char)', () => GPC.screen('G3RJM98NMQ'));
t('screen(short)', () => GPC.screen('G3RJM'));

t('toDMS', () => GPC.toDMS(43.650006, -79.380004));
t('toDMS(zero)', () => GPC.toDMS(0.0, 0.0));
t('toDMS(negzero)', () => GPC.toDMS(-0.0, -0.0));
t('toDMS(pole)', () => GPC.toDMS(90.0, 180.0));
t('fromDMS', () => GPC.fromDMS('43\u00b039\'00.02"N, 79\u00b022\'48.01"W'));
t('fromDMS(min60)', () => GPC.fromDMS('43\u00b060\'00.00"N, 79\u00b022\'48.01"W'));
t('fromDMS(sec60)', () => GPC.fromDMS('43\u00b039\'60.00"N, 79\u00b022\'48.01"W'));
t('fromDMS(junk)', () => GPC.fromDMS('hello'));
t('fromDMS(empty)', () => GPC.fromDMS(''));
t('toGeoURI', () => GPC.toGeoURI(43.650006, -79.380004));
t('toGeoURI(zero)', () => GPC.toGeoURI(0.0, 0.0));
t('fromGeoURI', () => GPC.fromGeoURI('geo:43.650006,-79.380004'));
t('fromGeoURI(crs)', () => GPC.fromGeoURI('geo:43.65,-79.38;crs=WGS84'));
t('fromGeoURI(badcrs)', () => GPC.fromGeoURI('geo:43.65,-79.38;crs=nad27'));
t('fromGeoURI(u)', () => GPC.fromGeoURI('geo:43.65,-79.38;u=35'));
t('fromGeoURI(alt)', () => GPC.fromGeoURI('geo:43.65,-79.38,120'));
t('fromGeoURI(junk)', () => GPC.fromGeoURI('http://x'));
t('fromGeoURI(empty)', () => GPC.fromGeoURI(''));

t('encodeAll(empty)', () => GPC.encodeAll([]).length);
t('decodeAll(empty)', () => GPC.decodeAll([]).length);
t('encodeAll(two)', () => GPC.encodeAll([[43.65, -79.38], [0.0, 0.0]]).join(' '));

t('withCheck', () => GPC.withCheck(C));
t('withCheck(reserved)', () => GPC.withCheck(X));

process.stdout.write(out.join('\n') + '\n');
